Multiuser and OpenId (#447)
actualbudget/docs: docs/advanced/oauth-auth.md

> # Authenticating With OpenID Provider

Reviewer: Is this plural or singular? You need to decide on either.

Author: I don't get it, where is this in plural?

Reviewer 2: I think it needs to be either:

```suggestion
# Authenticating With an OpenID Provider
```

or 

```suggestion
# Authenticating With OpenID Providers
```

> 

Author: added reasons

:::note
Client Version 24.10.0 and
Server Version 24.10.0 or higher are required for this feature.
:::

## Setup
:::info

> This feature requires use of [Actual Server](../config/)

Author: it does, you can use actual without the server

:::

To enable this feature, you can do it using configuration file at actual server, environment variables or you can use the UI

> ### Config using configuration file

Reviewer: Usually, each word in the headings is in upper case. You must be consistent across the document if you prefer not to have this.

Also, _configuration_ over _config_ makes for a more straightforward correct English sentence:

```suggestion
### Configuration Using a Configuration File
```

If your OpenId provider supports discovery, use the following configuration example:

```json title="config.json"
"openId": {
        "issuer": "URL for the OpenId Provider",
        "client_id": "client_id given by the provider",
        "client_secret": "client_secret given by the provider",
        "server_hostname": "your Actual Server URL (so the provider redirects you to this)"
    }
```

> If your OpenId provider does not provide discovery (like GitHub), you have to configure it manually:

Author: changed

```json title="config.json"
"openId": {
        "issuer": {
            "name": "Friendly name for the issuer",
            "authorization_endpoint": "Provider's authorize endpoint",
            "token_endpoint": "Provider's access token endpoint",
            "userinfo_endpoint": "Provider's user info endpoint"
        },
        "client_id": "client_id given by the provider",
        "client_secret": "client_secret given by the provider",
        "server_hostname": "your Actual Server URL (so the provider redirects you to this)"
    }
```

:::tip
This or the environment variables are the only ways to use GitHub as user provider.
:::

> ### Config using environment variables

Reviewer: Usually, each word in the headings is in upper case. If you prefer not to have this, you must be consistent across the document.

```suggestion
### Configuration Using Environment Variables
```

If your OpenId provider supports discovery, use the following variables:

- `ACTUAL_OPENID_DISCOVERY_URL`: URL for the OpenId Provider
- `ACTUAL_OPENID_CLIENT_ID`: client_id given by the provider
- `ACTUAL_OPENID_CLIENT_SECRET`: client_secret given by the provider
- `ACTUAL_OPENID_SERVER_HOSTNAME`: Your Actual Server URL (so the provider redirects you to this)

> If your OpenId provider does not provide discovery (like GitHub), use the following variables:

Reviewer: In this sentence _like GitHub_ points to that GitHub indeed provide discovery - but on line 45 it says GH does not support discovery.

Author: changed

- `ACTUAL_OPENID_AUTHORIZATION_ENDPOINT`: Provider's authorize endpoint
- `ACTUAL_OPENID_TOKEN_ENDPOINT`: Provider's access token endpoint
- `ACTUAL_OPENID_USERINFO_ENDPOINT`: Provider's user info endpoint
- `ACTUAL_OPENID_CLIENT_ID`: client_id given by the provider
- `ACTUAL_OPENID_CLIENT_SECRET`: client_secret given by the provider
- `ACTUAL_OPENID_SERVER_HOSTNAME`: Your Actual Server URL (so the provider redirects you to this)

> ### Configuring OpenID using UI

Reviewer: Usually, each word in the headings is in upper case. If you prefer not to have this, you must be consistent across the document.

```suggestion
### Configuring OpenID Using the UI
```

> #### Using previous versions Actual (with Actual Server enabled)

Reviewer: Usually, each word in the headings is in upper case. If you prefer not to have this, you must be consistent across the document.

Also - this sentence could be clearer. Should it be an _of_ inside there?

Why does the reference to previous versions of AB as the start of the document states a minium version number. This needs to be clearer.

Author: removed this section

> If you are using Actual Server for the past versions, you will need to enable OpenId Auth from the options:

Reviewer 3: Can you mention this is on the budget settings page. It may not be immediately obvious because OAuth isn't configured per budget it's per server (unless I'm misunderstanding)

Author: done

![](/static/img/oauth/start-using-options.png)

Once you click `Start using OpenID` a modal will be presented.

> #### Setup Actual Server for the first time

Reviewer: Usually, each word in the headings is in upper case. If you prefer not to have this, you must be consistent across the document.

Author: changed it to be consistent

> If you are starting to use Actual Server, during bootstrap, you can configure OpenID by click the button `Configure OpenID authentication instead (Advanced)`:

Reviewer: There is too much going on in this sentence simultaneously by having _during bootstrap_ in the middle. Please rewrite for clarity. 

_Clicking_ is grammatically more correct than using _click_.

Not everyone understands what _during bootstrap_ means in context. Could we use the words _initial setup_ instead?

```suggestion
During your initial setup of Actual, you can configure OpenID by clicking the button `Configure OpenID authentication instead (Advanced)`:
```

![](/static/img/oauth/welcome-button.png)

> ##### Configuration from options

Reviewer: Usually, each word in the headings is in upper case. You must be consistent across the document if you prefer not to have this.

Also, having just a screenshot without any explanation may be unclear to people.

![Configuration from options](/static/img/oauth/modal.png)

##### Configuration from bootstrap
![Configuration from bootstrap](/static/img/oauth/bootstrap.png)

##### Instructions
Fill all the required field for the selected provider.

> Some providers does not require all fields to be filled. 

Reviewer: This sentence is a bit incomplete. Add _out_ at the end. Also shorten it a bit:

```suggestion
Some providers only require some fields to be filled out.```

> For example passwordless.id does not have a `client_secret`.

Reviewer: I suggest using _As an example_ over _For example_ here.

```suggestion
As an example, the field _passwordless.id_ does not have a `client_secret`.
```

> After configuring all settings properly, click `OK`. The user will be redirected to the `login` page.

Reviewer: The sentence indicates that the reader is to perform a command (clicking on the OK button), and then we quickly move to the third person perspective.

```suggestion
When all settings are correctly filled out, click the 'OK' button, and you will be redirected to the `login` page.
```

:::warning

> There are some basic configuration checks when clicking `OK` but if you somehow type a wrong information and the data is saved, you will have to open the actual server database manually and update it. The name of the table is `auth`.

Reviewer: I suggest some small changes:

```suggestion
There are some basic configuration checks when clicking `OK`, but if you somehow type the wrong information and the data is saved, you must manually open the actual server database and update it. The table `auth` is the name of the table.
```

:::

#### Tested Providers
- Auth0
- Authentik
- GitHub
- Google Accounts
- Keycloak
- Microsoft Entra
- Passwordless.id

:::tip

> Each provider has different requirements, see Auth0 for example, you have to change for your own settings:

Reviewer: Too much going on in the same sentence. Split it up - my suggestion:

```suggestion
Each provider has different requirements. One example is Auth0, where you have to change the following:
```

![](/static/img/oauth/provider-requirement.png)
:::

#### After setup:

> You will be redirected to login page:

Reviewer: This is an incomplete sentence. My suggestion:

```suggestion
When setup is done, you will be redirected to the _login_ page:
```

![](/static/img/oauth/first-login.png)

### Next Step

> Once OpenID provider is setup. Setup [Multiuser](multi-user) 

Reviewer: Incomplete sentences. My suggestion:

```suggestion
Once you have configured your OpenID, you can continue to enable [Multiuser](multi-user) login. 
```

:::tip

> Configuring OpenID provider from options or bootstrap can be only done if your provider supports discovery, otherwise, use [file configuration](oauth-auth#config-using-configuration-file)

Reviewer: A lot is going on there at the same time. Alo _be only_ should be _only be_. _Bootstrap_ is, as stated above, a bit iffy to use.

I suggest the following:

```suggestion
Configuring the OpenID provider from options or during the initial setup for Actual can only be done if your provider supports discovery; otherwise, use [file configuration](oauth-auth#config-using-configuration-file)
```
